# Authors

*   Ismael Ruiz Garcia - ismael.ruiz.g@estudiantat.upc.edu
*   Tom Castro - tom.castro@estudiantat.upc.edu


# Exercise

We will be working with the same dataset and questions we dealt with in the previous lab, so that you have some reference to work on (and check if your results match).

We will use the [Adult UCI dataset](https://archive.ics.uci.edu/ml/datasets/adult) with a few modifications; download the following files: [description](./files/adults/adults.names), [data - part1](./files/adults/adults1.csv) [data - part2](./files/adults/adults2.csv).

Follow the instructions below and answer the questions. 

1. Create the two tables in the DB with the right data dypes. In order to use enumerated types in the table definition you will first need to create the type. For example:

```
CREATE TYPE mood AS ENUM ('sad', 'ok', 'happy');
CREATE TABLE person (
    name text,
    current_mood mood
);
INSERT INTO person VALUES ('Moe', 'happy');
SELECT * FROM person WHERE current_mood = 'happy';
 name | current_mood 
------+--------------
 Moe  | happy
```

Notice that each row/sample does not have an id. Instead of using an INTEGER data type for the id, we recommend you look into SERIAL.

2. Load the datasets. We should use this version of the copy_from command where we specify the name of the columns and the value for NULL in the file of origin. Notice how we are not specifying the id column, since that will be autogenerated by the DB. Caveat: once you have loaded the data, double check that the id/SERIAL column starts at 1. 
```
cursor.copy_from(f, 'adults1', columns=('age', 'workclass', ...), sep=',', null='?')
```
3. How many people under 18 years old have never worked? Of the never having worked people (all ages) is there any race bias (how many by race)? Is there any sex bias? - Note: no need to elaborate, just present the data in tables. 
4. Look at the hours per week of people with a paying job, by sex. Look at how many's income is above and below 50k. Compare and analyse. - Note: no need to elaborate, produce a single table that shows the data for both sex and salary.
5. How many people with college education do manual labour?
6. What is the minimum, mean and maximum capital gain and capital loss for every marital status?

## 1. Create the two tables in the DB with the right data dypes. 

First we are going to connect to our database. Then thanks to lab 1 we already know what the two databases are and what types should be each elements so we just have to create the two tables with the right types. But we also need to give an id to each table.

In [3]:
import psycopg2


try:
    conn = psycopg2.connect("dbname='postgres' user='postgres' host='localhost' password='pass1234'")
    print('Success connecting to the DB')
    cursor = conn.cursor()
except:
    print('I am unable to connect to the database')


Success connecting to the DB


In [ ]:
try:
    cursor.execute("""  CREATE TYPE workclass AS ENUM ('Private', 'Self-emp-not-inc', 'Self-emp-inc', 'Federal-gov', 'Local-gov', 'State-gov', 'Without-pay', 'Never-worked');
                        CREATE TYPE education as ENUM('Bachelors', 'Some-college', '11th', 'HS-grad', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', '9th', '7th-8th', '12th', 'Masters', '1st-4th', '10th', 'Doctorate', '5th-6th', 'Preschool');
                        CREATE TYPE "marital-status" as ENUM('Married-civ-spouse', 'Divorced', 'Never-married', 'Separated', 'Widowed', 'Married-spouse-absent', 'Married-AF-spouse');
                        CREATE TYPE occupation as ENUM('Tech-support', 'Craft-repair', 'Other-service', 'Sales', 'Exec-managerial', 'Prof-specialty', 'Handlers-cleaners', 'Machine-op-inspct', 'Adm-clerical', 'Farming-fishing', 'Transport-moving', 'Priv-house-serv', 'Protective-serv', 'Armed-Forces');
                        CREATE TABLE adults1 (
                            id SERIAL PRIMARY KEY,
                            age bigint,
                            workclass workclass,
                            fnlwgt bigint,
                            education education,
                            "education-num" bigint,
                            "marital-status" "marital-status",
                            occupation occupation
                        );
                        """)
    
    cursor.execute("""  CREATE TYPE relationship AS ENUM ('Wife', 'Own-child', 'Husband', 'Not-in-family', 'Other-relative', 'Unmarried');
                        CREATE TYPE race as ENUM('White', 'Asian-Pac-Islander', 'Amer-Indian-Eskimo', 'Other', 'Black');
                        CREATE TYPE sex as ENUM('Female', 'Male');
                        CREATE TYPE "native-country" as ENUM('United-States', 'Cambodia', 'England', 'Puerto-Rico', 'Canada', 'Germany', 'Outlying-US(Guam-USVI-etc)', 'India', 'Japan', 'Greece', 'South', 'China', 'Cuba', 'Iran', 'Honduras', 'Philippines', 'Italy', 'Poland', 'Jamaica', 'Vietnam', 'Mexico', 'Portugal', 'Ireland', 'France', 'Dominican-Republic', 'Laos', 'Ecuador', 'Taiwan', 'Haiti', 'Columbia', 'Hungary', 'Guatemala', 'Nicaragua', 'Scotland', 'Thailand', 'Yugoslavia', 'El-Salvador', 'Trinadad&Tobago', 'Peru', 'Hong', 'Holand-Netherlands');
                        CREATE TYPE income as ENUM('>50K', '<=50K');
                        CREATE TABLE adults2 (
                            id SERIAL PRIMARY KEY,
                            relationship relationship,
                            race race,
                            sex sex,
                            "capital-gain" bigint,
                            "capital-loss" bigint,
                            "hours-per-week" bigint,
                            "native-country" "native-country",
                            income income
                        );
                        """)
    
    conn.commit()
except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

## 2. Load the datasets. Once you have loaded the data, double check that the id/SERIAL column starts at 1. 

We simply need to open each adults.csv file and load their data in the adults table.

In [3]:
try:
    with open('./files/adults/adults1.csv', 'r') as f:
        cursor.copy_from(f, 'adults1', columns=('age','workclass','fnlwgt','education','education-num','marital-status','occupation'), sep=',', null='?')
        
    with open('./files/adults/adults2.csv', 'r') as g:
        cursor.copy_from(g, 'adults2', columns=('relationship','race','sex','capital-gain','capital-loss','hours-per-week','native-country','income'), sep=',', null='?')
        
    conn.commit()
except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

Now let's check the serial column. Let's start by getting the serial sequence.

In [4]:
try:    
    cursor.execute("""  SELECT pg_get_serial_sequence('adults1', 'id'), pg_get_serial_sequence('adults2', 'id');""")
    rows = cursor.fetchall()
    print(rows)
except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

[('public.adults1_id_seq', 'public.adults2_id_seq')]


Once we have the sequence we can check the first value 

In [5]:
try:
    cursor.execute("""  SELECT start_value FROM pg_sequences WHERE schemaname = 'public' AND sequencename IN ('adults1_id_seq','adults2_id_seq');""")
    rows = cursor.fetchall()
    print(rows)
except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

[(1,), (1,)]


We can see that the serial ID from both of the tables start with 1

## 3. How many people under 18 years old have never worked? Of the never having worked people (all ages) is there any race bias (how many by race)? Is there any sex bias? - Note: no need to elaborate, just present the data in tables. 

In [6]:
try:
    cursor.execute("""
        SELECT * FROM adults1
        WHERE age < 18 AND workclass = 'Never-worked';
    """)
    rows = cursor.fetchall()
    print(rows)

    cursor.execute("""
        SELECT COUNT(*) FROM adults1
        WHERE age < 18 AND workclass = 'Never-worked';
    """)
    count = cursor.fetchone()
    print("Total:", count[0])
except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

[(14773, 17, 'Never-worked', 237272, '10th', 6, 'Never-married', None)]
Total: 1


In [7]:
try:
    cursor.execute("""
        SELECT a2.race, COUNT(*) AS count_by_race
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        WHERE a1.workclass = 'Never-worked'
        GROUP BY a2.race
        ORDER BY count_by_race DESC;

    """)
    rows = cursor.fetchall()

    print("\nPeople who never worked by race:")
    print(f"{'Race':<25} {'Count':>10}")
    print("-" * 35)
    for race, count in rows:
        print(f"{race:<25} {count:>10,}")

    print('')

    cursor.execute("""
        SELECT a2.race, COUNT(*) AS count_by_race
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        GROUP BY a2.race
        ORDER BY count_by_race DESC;

    """)
    rows = cursor.fetchall()
    print("\nPeople by race:")
    print(f"{'Race':<25} {'Count':>10}")
    print("-" * 35)
    for race, count in rows:
        print(f"{race:<25} {count:>10,}")

except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)


People who never worked by race:
Race                           Count
-----------------------------------
White                              5
Black                              2


People by race:
Race                           Count
-----------------------------------
White                         27,816
Black                          3,124
Asian-Pac-Islander             1,039
Amer-Indian-Eskimo               311
Other                            271


In the never worked group, the ratio of whites to blacks is 2.5, whereas in the general dataset whites are over-represented 8.9x. Now, these numbers are small and cannot confirm the bias, however, it is a hint that warrants further investigation (although by looking at different datasets).


In [8]:
try:
    cursor.execute("""
        SELECT a2.sex, COUNT(*) AS count_by_race
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        WHERE a1.workclass = 'Never-worked'
        GROUP BY a2.sex
        ORDER BY count_by_race DESC;

    """)
    rows = cursor.fetchall()

    print("\nPeople who never worked by sex:")
    print(f"{'Race':<25} {'Count':>10}")
    print("-" * 35)
    for sex, count in rows:
        print(f"{sex:<25} {count:>10,}")

    print('')

    cursor.execute("""
        SELECT a2.sex, COUNT(*) AS count_by_race
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        GROUP BY a2.sex
        ORDER BY count_by_race DESC;

    """)
    rows = cursor.fetchall()
    print("\nPeople by sex:")
    print(f"{'Sex':<25} {'Count':>10}")
    print("-" * 35)
    for sex, count in rows:
        print(f"{sex:<25} {count:>10,}")

except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)


People who never worked by sex:
Race                           Count
-----------------------------------
Male                               5
Female                             2


People by sex:
Sex                            Count
-----------------------------------
Male                          21,790
Female                        10,771


We cannot conclude there is sex bias, since the proportions are similar and most importantly, the number of samples is too small.


## 4. Look at the hours per week of people with a paying job, by sex. Look at how many's income is above and below 50k. Compare and analyse. - Note: no need to elaborate, produce a single table that shows the data for both sex and salary.

In [9]:
try:
    cursor.execute("""
        SELECT a2.sex,
        COUNT(*),
        ROUND(AVG(a2."hours-per-week"), 2),
        MIN(a2."hours-per-week"),
        MAX(a2."hours-per-week"),
        PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY a2."hours-per-week") AS median_hours
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        WHERE a1.workclass NOT IN ('Without-pay', 'Never-worked')
        GROUP BY a2.sex
        ORDER BY a2.sex;
    """)
    rows = cursor.fetchall()
    print("\nHours per week of people with a paying job, by sex:")
    print(f"{'Sex':<10} {'Count':>10} {'Avg Hours':>12} {'Min':>6} {'Median':>8} {'Max':>6}")
    print("-" * 60)
    for sex, count, avg, min_h, max_h, median in rows:
        print(f"{sex:<10} {count:>10,} {float(avg):>12.2f} {min_h:>6} {median:>8} {max_h:>6}")


    cursor.execute("""
        SELECT a2.sex,
        a2.income,
        COUNT(*),
        ROUND(AVG(a2."hours-per-week"), 2),
        MIN(a2."hours-per-week"),
        MAX(a2."hours-per-week"),
        PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY a2."hours-per-week") AS median_hours
        FROM adults1 a1
        JOIN adults2 a2 ON a1.id = a2.id
        WHERE a1.workclass NOT IN ('Without-pay', 'Never-worked')
        GROUP BY a2.sex, a2.income
        ORDER BY a2.sex, a2.income;
    """)

    rows = cursor.fetchall()
    print("\nIncome distribution of people with a paying job, by sex:")
    print(f"{'Sex':<10} {'Income':<6} {'Count':>10} {'Avg Hours':>12} {'Min':>6} {'Median':>8} {'Max':>6}")
    print("-" * 70)
    for sex, income, count, avg, min_h, max_h, median in rows:
        print(f"{sex:<10} {income:<6} {count:>10,} {float(avg):>12.2f} {min_h:>6}  {median:>8} {max_h:>6}")

except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)


Hours per week of people with a paying job, by sex:
Sex             Count    Avg Hours    Min   Median    Max
------------------------------------------------------------
Female          9,925        36.96      1     40.0     99
Male           20,779        42.86      1     40.0     99

Income distribution of people with a paying job, by sex:
Sex        Income      Count    Avg Hours    Min   Median    Max
----------------------------------------------------------------------
Female     >50K        1,127        40.82      2      40.0     99
Female     <=50K       8,798        36.46      1      40.0     99
Male       >50K        6,523        46.55      1      45.0     99
Male       <=50K      14,256        41.17      1      40.0     99


When examining the hours worked per week among people with a paying job, there is a clear difference between men and women.
On average, men work about 42.9 hours per week, while women work around 37 hours, meaning that men work roughly 6 hours more per week on average.

When we analyze income distribution, the gender gap becomes even more evident.
Among people with paying jobs:

* Only 11.4% of women earn more than 50K, while 88.6% earn 50K or less.
* In contrast, 31.4% of men earn more than 50K, and 68.6% earn 50K or less.

There is also a positive correlation between hours worked and higher income: both men and women who earn more than 50K work longer hours on average (about 46.6 hours for men and 40.8 hours for women) compared to those earning 50K or less.

Overall, these results highlight a double disparity: men tend to work longer hours and are also much more likely to hold higher-paying positions, which together contribute to a significant gender income gap among people with paying jobs.

## 5. How many people with college education do manual labour?

In [17]:
try:

    cursor.execute("""
        SELECT
            COUNT(*) FILTER (WHERE occupation IN ('Tech-support', 'Sales', 'Exec-managerial', 'Prof-specialty', 'Adm-clerical')) AS desk,
            COUNT(*) FILTER (WHERE occupation NOT IN ('Tech-support', 'Sales', 'Exec-managerial', 'Prof-specialty', 'Adm-clerical')) AS manual
        FROM adults1
        WHERE occupation IS NOT NULL
          AND "education-num" IS NOT NULL
          AND workclass IS NOT NULL
          AND "education-num" >= 10;
        """)
    rows = cursor.fetchall()
    colnames = [desc[0] for desc in cursor.description]

    print(f"{'occupation_type':<12} | {'count_people':>12}")
    print("-" * 28)
    for col, val in zip(colnames, rows[0]):
        print(f"{col:<12} | {val:>12}")

except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

occupation_type | count_people
----------------------------
desk         |        12207
manual       |         4722


The column `education` contains the categorical values that describe a person’s level of studies, while the column `education-num` provides a corresponding numerical scale, where 1 represents the most basic level of education and 16 corresponds to the highest, a doctorate. Based on this information, we can define college education as those cases where `education-num` >= 10.

4722 people with some college education do manual jobs.

## 6. What is the minimum, mean and maximum capital gain and capital loss for every marital status?

In [41]:
try:

    cursor.execute("""
            SELECT
                a1."marital-status"               AS marital_status,
                MIN(a2."capital-gain")            AS min_capital_gain,
                AVG(a2."capital-gain")            AS mean_capital_gain,
                MAX(a2."capital-gain")            AS max_capital_gain,
                MIN(a2."capital-loss")            AS min_capital_loss,
                AVG(a2."capital-loss")            AS mean_capital_loss,
                MAX(a2."capital-loss")            AS max_capital_loss
            FROM adults1 a1
            JOIN adults2 a2
            ON a1.id = a2.id   
            GROUP BY a1."marital-status"
            ORDER BY a1."marital-status";
        """)


    rows = cursor.fetchall()
    colnames = [desc[0] for desc in cursor.description]

    # En-tête du tableau
    print(f"{'':<50}{'Capital_Gain':>10}     |     {'Capital_Loss':>10}")
    print(" "*26+"-" * 84)
    print(f"{'marital_status':<25} | {'min_gain':>10} | {'mean_gain':>12} | {'max_gain':>10} | {'min_loss':>10} | {'mean_loss':>12} | {'max_loss':>10}")
    print("-" * 110)

    # Contenu du tableau
    for row in rows:
        marital_status, min_gain, mean_gain, max_gain, min_loss, mean_loss, max_loss = row
        print(f"{marital_status:<25} | "
              f"{min_gain:>10.0f} | "
              f"{mean_gain:>12.2f} | "
              f"{max_gain:>10.0f} | "
              f"{min_loss:>10.0f} | "
              f"{mean_loss:>12.2f} | "
              f"{max_loss:>10.0f}")


except Exception as e:
    # if the transaction aborts we will need to rollback
    cursor.execute("ROLLBACK")
    print(e)

                                                  Capital_Gain     |     Capital_Loss
                          ------------------------------------------------------------------------------------
marital_status            |   min_gain |    mean_gain |   max_gain |   min_loss |    mean_loss |   max_loss
--------------------------------------------------------------------------------------------------------------
Married-civ-spouse        |          0 |      1764.86 |      99999 |          0 |       122.44 |       2603
Divorced                  |          0 |       728.41 |      99999 |          0 |        64.60 |       3900
Never-married             |          0 |       376.59 |      99999 |          0 |        52.35 |       3683
Separated                 |          0 |       535.57 |      99999 |          0 |        58.11 |       3900
Widowed                   |          0 |       571.07 |      99999 |          0 |        79.67 |       4356
Married-spouse-absent     |          0 |    

> Capital Gain

We can see that for all marital-status groups, the **minimum capital gain is 0**, meaning that at least one individual in each group did not receive any gains.

The **average capital gain** is higher for `Married-civ-spouse` and `Divorced` individuals. The higher mean for `Married-civ-spouse` could be due to greater financial stability in dual-income households, allowing for larger or more frequent investments.

The **maximum capital gains** are extremely high (**99999**) for most groups, highlighting the presence of outliers. These extreme values increase the standard deviation, making the spread of the data appear larger. `Married-AF-spouse` is an exception, with a lower maximum gain (**7298**) and smaller standard deviation, likely because the sample size for this group is very small.



---



> Capital Loss

Similarly, the **minimum capital loss is 0** for all groups, indicating that at least one individual in each group did not lose money.

The **average capital loss** is highest among `Married-civ-spouse`, which may suggest that individuals with higher financial stability are willing to take greater risks, resulting in larger losses. Most other groups have a mean loss around 50–60. The `Married-AF-spouse` group shows **0** average loss, likely due to the very small number of observations.

The **maximum capital loss** is around **3900** for most groups, again reflecting the presence of extreme cases. `Married-AF-spouse` reports no losses, which is probably because the sample size is too small to include any cases of loss.